orb-ui/src/engine/animations/alert.rs

```diff
@@ -0,0 +1,169 @@
+use crate::engine::rgb::Argb;
+use crate::engine::{Animation, AnimationState};
+use std::any::Any;
+use std::f64::consts::PI;
+
+/// Alert animation.
+/// Blink following a pattern describing the durations for each edge.
+/// The first edge can be set with LEDs on or off.
+/// The pattern contains the consecutive edges duration.
+/// Smooth transitions can optionally be set to start a sine wave to transition between edges.
+
+/// Pattern contains the consecutive edges duration
+/// Edge starts high/on if `active_at_start` is `true`, off otherwise
+/// Example: vec![0.0, 0.3, 0.2, 0.3]
+///          0.0 to 0.3 `active_at_start`, 0.3 to 0.5 `!active_at_start`, 0.5 to 0.8 `active_at_start`
+///          ends up with the `active_at_start` edge
+type Pattern = Vec<f64>;
+
+/// Blink following a pattern describing the durations for each
+/// edge. The first edge can be set with LEDs on or off.
+pub struct Alert<const N: usize> {
+    /// Currently rendered color.
+    current_solid_color: Argb,
+    target_color: Argb,
+    /// pattern contains the consecutive edges duration
+    pattern: Pattern,
+    /// allows to start a sine wave to transition between edges.
+    /// The color will start to change from the current color to the target color by the time given
+    /// between `smooth_transitions[i]` and `pattern[i+1]`.
+    /// example with pattern vector above, and `smooth_transitions`:
+    ///         Some(vec![0.1, 0.1, 0.1])
+    ///         t=0.0 to t=0.1 `active_at_start`, t=0.1 to t=0.3 transition to `!active_at_start`, etc.
+    smooth_transitions: Option<Vec<f64>>,
+    /// time in animation
+    phase: f64,
+    /// first edge from pattern\[0\] to pattern\[1\] has LEDs on
+    active_at_start: bool,
+}
+
+impl<const N: usize> Alert<N> {
+    /// Creates a new [`Alert`].
+    #[must_use]
+    pub fn new(
+        color: Argb,
+        pattern: Vec<f64>,
+        smooth_transitions: Option<Vec<f64>>,
+        active_at_start: bool,
+    ) -> Self {
+        Self {
+            target_color: color,
+            current_solid_color: if active_at_start { color } else { Argb::OFF },
+            smooth_transitions,
+            pattern,
+            phase: 0.0,
+            active_at_start,
+        }
+    }
+}
+
+impl<const N: usize> Animation for Alert<N> {
+    type Frame = [Argb; N];
+
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn as_any_mut(&mut self) -> &mut dyn Any {
```

Comment:
have a look at the Trait and usages in the code.

Reply:
I think what I mean is: It makes more sense to use `dyn Animation` (which is already what is stored in the animation stack) than `dyn Any`. If you want the ability to downcast animations, IMO it would be better to create a `downcast_ref()` and `downcast_mut()` function on the Animation trait, which in its default impl, calls std::any::Any::downcast_ref to do the downcast.

The reason this is prefereable over using dyn Any is that dyn Animation is a more narrow type, and that it makes the code a bit easier to understand.

Reply:
tbh I didn't write that piece of code. it was taken from orb-core.
I am not really sure how to implement this, can you propose a PR if you think it deserves it?
thanks